design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,70 @@
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as well as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy > currentHealthy, we can disable fencing on the nodes the MDB points to.
+
+# Design:
+
+We will implement two controllers `machinedisruptionbudget-controller` and the `machine-controller` to be implemented through the controller pattern describere [here](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg#hdr-Controller_Writing_Tips). Each controller watches a set of object kinds and reconciles one.
+
+## machinedisruptionbudget-controller
+This controller watches ceph PGs and CephClusters. We will ensure the reconciler is enqueued every 60s. It ensures that each CephCluster has a MDB created, and the MDB's value of maxUnvailable reflects the health of the Ceph Cluster's PGs.
+If all PGs are clean, maxUnavailable = 1.
+else, maxUnavailable = 0.
+
+We can share a ceph health cache with the other controller-runtime reconcilers that have to watch the PG "cleanliness".
+
+The MDB will target `Machine`s selected by a label maintainted by the `machine-controller`. The label is `fencegroup.rook.io/<cluster-name>`. 
+
+## machine-controller
+This controller watches OSDs and `Machine`s. It ensures that each `Machine` with OSDs from a `CephCluster` have the label `fencegroup.rook.io/<cluster-name>`, and those that do not have running OSDs do not have label.
+
+This will ensure that no `Machine` without running OSDs will be protected by the MDB.
+
+## Assumptions:
+- We assume that the controllers will be able to reconcile multiple times in < 5 minutes as we know that fencing will happen only after a configurable timeout. The default timeout is 5 minutes.
+  This is important because the MDB must be reconciled based on an accurate ceph health state in that time.
+
+
+## Example Flows:
+
+**Node needs to be fenced, the OSDs on the node are down too**
+
+ - Node has NotReady condition.
+ - Some Ceph PGs are not active+clean.
+ - machinedisruptionbudget-controller sets maxUnavailable to 0 on the MachineDisruptionBudget.
+ - MachineHealthCheck sees NotReady and attempts to fence after 5 minutes, but can't due to MDB
+ - machine-controller notices all OSDs on the affected node are down and removes the node from the MDB.
+ - MDB no longer covers the affected node, and MachineHealthCheck fences it.
+
+**Node needs to be fenced, but the OSDs on the node are up**
+
+ - Node has Notready condition.
+ - Ceph PGs are all active+clean so maxUnavailable remains 1 on the MDB.
+ - MachineHealthCheck fences the Node.
+ - Ceph resources on the node go down.
+ - Some Ceph PGs are now not active+clean.
+ - machinedisruptionbudget-controller sets maxUnavailble to 0 on the MachineDisruptionBudget.
+ - If another labeled node needs to be fenced, it will only happen after the Ceph PGs become active+clean again when the OSDs are rescheduled and backfilled.
```

Comment:
@rohantmp bottom line is that if PG state is unhealthy fencing is blocked at the machine level - but fence requests for machines without an OCS label will be able to carry on regardless of ceph state.

Reply:
Yup.

Reply:
I'd update the doc to capture this point to avoid confusion - maybe under the goals section.

Reply:
+1

Reply:
What happens if another node fails before the originally failed node comes up?
It would be safe to fence the second node once the original node becomes available and starts its OSDs, however I don't think that would happen based on this design (since there would be no point at which all Ceph PGs are active+clean)